### PR TITLE
feat(float): open float relative to mouse

### PR DESCRIPTION
**Pull request by seblyng**
After #19481 I have thought a bit about having hover from lsp opening when the mouse hovers over like in other IDE's. This has previously been difficult because of no option to open a window relative to the mouse.

This PR introduces another option to the `relative` key in `nvim_open_win()`. It allows it to open a floating window relative to the mouse.

With this PR it should be possible to override the handler and do something similar to this https://github.com/neovim/neovim/pull/19481#issuecomment-1193248674 to have hover information displayed from the mouse.

Let me know what you think!

PS: I am not too sure about how it works in `window.c` with the floating window options, but it seems to work okay. Let me know if you see anything that may be wrong!

If anyone is interested in testing this with the hover on mouse, I'll paste a snippet for anyone to try out

<details>
<summary>Mouse hover snippet</summary>

```lua
local util = require('vim.lsp.util')

local function make_position_param(window, offset_encoding)
    window = window or 0
    local buf = vim.api.nvim_win_get_buf(window)
    local row, col

    local mouse = vim.fn.getmousepos()
    row = mouse.line
    col = mouse.column

    offset_encoding = offset_encoding or util._get_offset_encoding(buf)
    row = row - 1
    local line = vim.api.nvim_buf_get_lines(buf, row, row + 1, true)[1]
    if not line then
        return { line = 0, character = 0 }
    end
    if #line < col then
        return { line = 0, character = 0 }
    end

    col = util._str_utfindex_enc(line, col, offset_encoding)

    return { line = row, character = col }
end

local make_params = function(window, offset_encoding)
    window = window or 0
    local buf = vim.api.nvim_win_get_buf(window)
    offset_encoding = offset_encoding or util._get_offset_encoding(buf)
    return {
        textDocument = util.make_text_document_params(buf),
        position = make_position_param(window, offset_encoding),
    }
end

local hover_timer = nil
vim.o.mousemoveevent = true

vim.keymap.set({ '', 'i' }, '<MouseMove>', function()
    if hover_timer then
        hover_timer:close()
    end
    hover_timer = vim.defer_fn(function()
        hover_timer = nil
        local params = make_params()
        vim.lsp.buf_request(
            0,
            'textDocument/hover',
            params,
            vim.lsp.with(vim.lsp.handlers.hover, {
                silent = true,
                focusable = false,
                relative = 'mouse',
            })
        )
    end, 500)
    return '<MouseMove>'
end, { expr = true })
```

</details>

### CODE REVIEW

**Review by justinmk — APPROVED** [review]
makes sense to me, we didn't have `getmousepos` when floating windows were introduced IIRC.

just needs a rebase

**Comment by seblj** [comment]
> just needs a rebase

done

File: src/nvim/window.c [hunk 1]
```diff
@@ -798,6 +798,16 @@ void win_config_float(win_T *wp, FloatConfig fconfig)
     fconfig.row += curwin->w_wrow;
     fconfig.col += curwin->w_wcol;
     fconfig.window = curwin->handle;
+  } else if (fconfig.relative == kFloatRelativeMouse) {
+    int row = mouse_row, col = mouse_col, grid = mouse_grid;
+    win_T *mouse_win = mouse_find_win(&grid, &row, &col);
+    if (mouse_win == NULL) {
+      return;
```
Comment (zeertzjq) [inline src/nvim/window.c:805]: This `return` doesn't look safe. When no window can be found I think it can be relative to editor instead?

Comment (seblj) [reply src/nvim/window.c]: done

**Comment by justinmk** [comment]
Thanks! (And thanks @zeertzjq for catching that bug).

I added context to the commit message from this PR description. In the future please include that (ideally in problem/solution format).

**Pull request merged**